packages/create-amplify/src/package-manager-controller/package_manager_controller_factory.ts

```diff
@@ -0,0 +1,157 @@
+      executable: 'npm',
+      binaryRunner: 'npx',
+      installCommand: 'install',
+      lockFile: 'package-lock.json',
+      initDefault: ['init', '--yes'],
+    },
+    'yarn-classic': {
+      name: 'yarn-classic',
+      executable: 'yarn',
+      binaryRunner: 'yarn',
+      installCommand: 'add',
+      lockFile: 'yarn.lock',
+      initDefault: ['init', '--yes'],
+    },
+    'yarn-modern': {
+      name: 'yarn-modern',
+      executable: 'yarn',
+      binaryRunner: 'yarn',
+      installCommand: 'add',
+      lockFile: 'yarn.lock',
+      initDefault: ['init', '--yes'],
+    },
+    pnpm: {
+      name: 'pnpm',
+      executable: 'pnpm',
+      binaryRunner: 'pnpm',
+      installCommand: 'add',
+      lockFile: 'pnpm-lock.yaml',
+      initDefault: ['init'],
+    },
+  };
+
+  /**
+   * Check if a package.json file exists in projectRoot
+   */
+  protected packageJsonExists = (projectRoot: string): boolean => {
+    return this.existsSync(path.resolve(projectRoot, 'package.json'));
+  };
+
+  /**
+   * getPackageManager
+   */
+  getPackageManager() {
+    if (!process.env.npm_config_user_agent) {
+      logger.warn('Could not determine package manager, defaulting to npm');
+      return this.packageManagers['npm'];
+    }
+
+    const userAgent = process.env.npm_config_user_agent;
+    const packageManagerAndVersion = userAgent.split(' ')[0];
+    const packageManagerName = packageManagerAndVersion.split('/')[0];
+
+    if (packageManagerName === 'yarn') {
+      const yarnMajorVersion = packageManagerAndVersion
+        .split('/')[1]
+        .split('.')[0];
+      const yarnName: PackageManagerName = `${packageManagerName}-${
+        yarnMajorVersion === '1' ? 'classic' : 'modern'
+      }`;
+      return this.packageManagers[yarnName];
+    }
+    return this.packageManagers[packageManagerName as PackageManagerName];
+  }
+
+  /**
+   * getPackageManagerController
+   */
+  getPackageManagerController(projectRoot: string) {
+    if (!this.packageManagerName) {
+      this.packageManagerName = this.getPackageManager().name;
+    } else if (this.packageManagerName !== this.getPackageManager().name) {
+      throw new Error("There's a conflicts of the package manager name.");
+    }
+    switch (this.packageManagerName) {
+      case 'npm':
+        return new NpmPackageManagerController(projectRoot);
+      case 'pnpm':
+        return new PnpmPackageManagerController(projectRoot);
+      case 'yarn-classic':
+        return new YarnClassicPackageManagerController(projectRoot);
+      case 'yarn-modern':
+        return new YarnModernPackageManagerController(projectRoot);
+      default:
+        return new NpmPackageManagerController(projectRoot);
+    }
+  }
+
+  /**
+   * getProjectInitializer
+   */
+  getProjectInitializer(projectRoot: string) {
+    if (!this.packageManagerName) {
+      this.packageManagerName = this.getPackageManager().name;
+    } else if (this.packageManagerName !== this.getPackageManager().name) {
```

Comment:
How could this happen? Can we structure the code in a way that this is impossible?

Reply:
removed